src/util/log.h

```diff
@@ -14,12 +14,47 @@
 #define ASSERT(pass, msg, ...) _smek_assert(__FILE__, __LINE__, __func__, pass, STR(pass), msg, ##__VA_ARGS__)
 #define CHECK(pass, msg, ...)  _smek_check(__FILE__, __LINE__, __func__, pass, STR(pass), msg, ##__VA_ARGS__)
 
-#define ASSERT_EQ(LHS, RHS) ASSERT((LHS) == (RHS), STR(LHS) " ({}) and " STR(RHS) " ({}) should be equal", (LHS), (RHS));
-#define ASSERT_NE(LHS, RHS) ASSERT((LHS) != (RHS), STR(LHS) " ({}) and " STR(RHS) " ({}) should not be equal", (LHS), (RHS));
-#define ASSERT_LT(LHS, RHS) ASSERT((LHS) < (RHS), STR(LHS) " ({}) should be less than " STR(RHS) " ({})", (LHS), (RHS));
-#define ASSERT_GT(LHS, RHS) ASSERT((LHS) > (RHS), STR(LHS) " ({}) should be greater than " STR(RHS) " ({})", (LHS), (RHS));
-#define ASSERT_LE(LHS, RHS) ASSERT((LHS) <= (RHS), STR(LHS) " ({}) should be less than or equal to " STR(RHS) " ({})", (LHS), (RHS));
-#define ASSERT_GE(LHS, RHS) ASSERT((LHS) >= (RHS), STR(LHS) " ({}) should be greater than or equal to " STR(RHS) " ({})", (LHS), (RHS));
+#define ASSERT_EQ(LHS, RHS)                                                                           \
+    do {                                                                                              \
+        auto lhs = LHS;                                                                               \
+        decltype(lhs) rhs = RHS;                                                                      \
```

Comment:
We could swap this to a "auto", but it might cause extra casting and behave strangely. Here I assume `lhs` and `rhs` can be cast to eachother. But I don't know what's best.